crates/vm/src/system/memory/online.rs

```diff
@@ -515,133 +443,271 @@ impl<F: PrimeField32> TracingMemory<F> {
+         *
+         * The way adapter records work is:
+         * - Every time we have an access, it has two values:
+         *   - The size of the access,
+         *   - The size of the subsegments we may want to split it into. At the time of the
+         *     access, we don't know yet if we want to split it afterwards. This small size is
+         *     usually equal to the `align` for this address space, but it can differ, e.g. if
+         *     we want to split `align=4` into final segments, which have size `1` in the
+         *     volatile memory interface.
+         * - We add one record to all the adapters with sizes from `(lowest_size, block_size]`.
+         * - When we want to mark an access to be split, we only modify the highest record (the
+         *   one with the largest size), which is considered the "master" record. The
+         *   corresponding metadatas will have the offset of this record in its arena.
+         * - Before finalization, we call `prepare_to_finalize` function that propagates the
+         *   split flag to all the smaller records for this access.
+         */
         let num_segs = BLOCK_SIZE / align;
 
         let begin = pointer / align;
-        let end = begin + BLOCK_SIZE / align;
-
-        let mut prev_ts = INITIAL_TIMESTAMP;
-        let mut block_timestamps = vec![INITIAL_TIMESTAMP; num_segs];
-        let mut cur_ptr = begin;
-        let need_to_merge = loop {
-            if cur_ptr >= end {
-                break true;
-            }
-            // SAFETY:
-            // - alignment is automatic since we multiply by `size_of::<AccessMetadata>()`
-            let mut current_metadata = unsafe {
-                self.meta
-                    .get_unchecked(address_space)
-                    .read::<AccessMetadata>(cur_ptr * size_of::<AccessMetadata>())
-            };
-            if current_metadata.block_size == BLOCK_SIZE as u32 && cur_ptr + num_segs == end {
-                // We do not have to do anything
-                prev_ts = current_metadata.timestamp;
-                break false;
-            } else if current_metadata.block_size == 0 {
-                // Initialize
-                if self.initial_block_size < align {
-                    // Only happens in volatile, so empty initial memory
-                    self.set_meta_block(
-                        address_space,
-                        cur_ptr * align,
-                        align,
-                        align,
-                        INITIAL_TIMESTAMP,
-                    );
-                    current_metadata = AccessMetadata::new(INITIAL_TIMESTAMP, align as u32);
+
+        let first_meta = unsafe {
+            self.meta[address_space].read::<AccessMetadata>(begin * size_of::<AccessMetadata>())
+        };
+        let mut need_to_merge = first_meta.block_size != BLOCK_SIZE as u32;
+        let existing_metadatas = (0..num_segs)
```

Comment:
feels like this does a collect always, I still want use to break very early if `BLOCK_SIZE = align` and there's nothing to do